src/util-ja3.c

```diff
@@ -0,0 +1,290 @@
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Mats Klepsland <mats.klepsland@gmail.com>
+ *
+ * Functions used to generate JA3 fingerprint.
+ */
+
+#include "suricata-common.h"
+
+#include "util-ja3.h"
+
+#define MD5_STRING_LENGTH 33
+
+/**
+ * \brief Allocate new buffer.
+ *
+ * \return pointer to buffer on success.
+ * \return NULL on failure.
+ */
+JA3Buffer *Ja3BufferInit(void)
+{
+    JA3Buffer *buffer = SCCalloc(1, sizeof(JA3Buffer));
+    if (buffer == NULL) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Error allocating memory for JA3 buffer");
+        return NULL;
+    }
+
+    return buffer;
+}
+
+/**
+ * \brief Free allocated buffer.
+ *
+ * \param buffer The buffer to free.
+ */
+void Ja3BufferFree(JA3Buffer *buffer)
+{
+    if (buffer == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "Buffer should not be NULL");
+        return;
+    }
+
+    if (buffer->data != NULL) {
+        SCFree(buffer->data);
+    }
+
+    SCFree(buffer);
+}
+
+/**
+ * \internal
+ * \brief Resize buffer if it is full.
+ *
+ * \param buffer The buffer.
+ * \param len    The length of the data that should fit into the buffer.
+ *
+ * \retval 0 on success.
+ * \retval -1 on failure.
+ */
+static int Ja3BufferResizeIfFull(JA3Buffer *buffer, uint32_t len)
+{
+    if (buffer == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "Buffer should not be empty");
+        return -1;
+    }
+
+    if (len == 0) {
+        return 0;
+    }
+
+    while (buffer->used + len + 2 > buffer->size)
+    {
+        buffer->size *= 2;
+        char *tmp = SCRealloc(buffer->data, buffer->size * sizeof(char));
+        if (tmp == NULL) {
+            SCLogError(SC_ERR_MEM_ALLOC, "Error resizing JA3 buffer");
+            return -1;
+        }
+        buffer->data = tmp;
+    }
+
+    return 0;
+}
+
+/**
+ * \brief Append buffer to buffer.
+ *
+ * Append the second buffer to the first and then free it.
+ *
+ * \param buffer1 The first buffer.
+ * \param buffer2 The second buffer.
+ *
+ * \retval 0 on success.
+ * \retval -1 on failure.
+ */
+int Ja3BufferAppendBuffer(JA3Buffer *buffer1, JA3Buffer *buffer2)
+{
+    if (buffer1 == NULL || buffer2 == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "Buffers should not be NULL");
+        return -1;
+    }
+
+    /* If buffer1 contains no data, then we just copy the second buffer
+       instead of appending its data. */
+    if (buffer1->data == NULL) {
+        Ja3BufferFree(buffer1);
+        *buffer1 = *buffer2;
```

Comment:
then here it's written to, that looks like use after free

Reply:
This is by design, buffer1 is set to the value of buffer2 if buffer1 is empty, instead of copying buffer2. Should this be changed?

Reply:
Yes. The memory at 'buffer1' is freed. The line *buffer1 = *buffer2 copies the content of buffer2 (so the data at this memory location, not the pointer address itself) to the memory buffer1 points to. But this was just freed. So this is use after free.

Reply:
Shit, that's true. You are right, as usual!